contracts/PetPark.sol

```diff
@@ -1,7 +1,110 @@
 //SPDX-License-Identifier: Unlicense
 pragma solidity ^0.8.0;
 
+import "@openzeppelin/contracts/access/Ownable.sol";
 
-contract PetPark {
+contract PetPark is Ownable {
+    // TODO: Don't use storage?
+    uint8 constant ANIMAL_TYPE_FISH = 1;
+    uint8 constant ANIMAL_TYPE_CAT = 2;
+    uint8 constant ANIMAL_TYPE_RABBIT = 4;
+    uint8 constant ANIMAL_TYPE_DOG = 3;
+    uint8 constant ANIMAL_TYPE_PARROT = 5;
+    //enum AnimalType{ NONE, FISH, CAT, DOG, RABBIT, PARROT }
 
-}
\ No newline at end of file
+    modifier validRange(
+        uint256 val,
+        uint256 min,
+        uint256 max
+    ) {
+        require(val >= min && val <= max, "Invalid animal type");
+        _;
+    }
+
+    struct Borrow {
+        uint8 age;
+        uint8 gender;
+        uint8 animalType;
+    }
+
+    mapping(uint8 => uint256) private _animalCounts;
+    mapping(address => Borrow) private _ownerBorrow;
+
+    event Added(uint8 animalType, uint256 count);
+    event Borrowed(uint8 animalType);
+
+    function animalCounts(uint8 _animalType)
+        external
+        view
+        returns (uint256 animalCount)
+    {
+        return _animalCounts[_animalType];
+    }
+
+    function add(uint8 _animalType, uint256 _count)
+        external
+        onlyOwner
+        validRange(_animalType, 1, 5)
+    {
+        //require(_animalType > 0 && _animalType <= 5, "Invalid animal type");
+        _animalCounts[_animalType] += _count;
+        emit Added(_animalType, _count);
+    }
+
+    function borrow(
+        uint8 _age,
+        uint8 _gender,
+        uint8 _animalType
+    ) external validRange(_animalType, 1, 5) {
+        require(_age > 0, "Invalid Age");
+        if (_ownerBorrow[msg.sender].animalType > 0) {
```

Comment:
Since `_ownerBorrow[msg.sender]` is being used multiple times it can be stored in a single memory variable. Later in the course you will learn that it optimises for gas usage